### PR TITLE
Generating Synthetic Brain Scans with StyleGAN (COMP3710)

### PR DESCRIPTION
**Pull request by MiNat015**
This pull request implements a StyleGAN2-based architecture for generating synthetic brain MRI scans using the ADNI dataset as part of COMP3710. The implementation focuses on demonstrating the potential of deep learning in medical imaging synthesis while maintaining computational efficiency.

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: StyleGAN on ADNI ---- Hard Difficulty

1. Algorithm solves the problem: (3.5 Marks)
Somewhat reasonably clear generated images;
Resized to 64x64 would lose much information and make it hard to see how clear the generated images are  -0.5;
No embedding on style space (and a brief justification) reported on README.md -1

2. Algorithm demonstrated and functions: (3 Marks)

4. Design: (1 Mark)
Modularized design

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:
- It's not really necessary to resize from 256x240 to 64x64, computation saved does not worth information lost and performance jeopardized;
- Nice training documentation with justification on hyperparameter tunings, as well as phased LR;
- I agree with your statements under Results and Future Improvements, higher resolution could make a difference;
- I see you have both the function of implementing embedding in `dataset.py` and an embedding plot, but the function was not used anywhere in the code and the embedding plot was not shown in the README.md --- it would be good if you can clarify. Also, the embedding plot should be on the style space, I can't tell if you did so since no usage of the function present;
- Good referencing practice.

**Comment by yexincheng** [comment]
**This is an initial inspection, no action is required at this point**
1. **Solves problem**: Nice documentation and good introduction of the network. The generated images are blurry as the image resolution is 64 * 64; some are ok but some miss the structures. The loss plot was provided. Nice records of experiment steps. UMAP or T-SNE result was not presented, but the related functions were implemented.
2. **Implementation**: Data augmentation implemented. Nice and well structured.
3. **Design**: Good
4. **Comment**: Basic comments, could have more detailed comments.
5. **Difficulty**: Hard

**Note**
Student ID number is required. Is the random seed (manual_seed = 46915474) your ID number?

**Comment by MiNat015** [comment]
> **This is an initial inspection, no action is required at this point**
> 
> 1. **Solves problem**: Nice documentation and good introduction of the network. The generated images are blurry as the image resolution is 64 * 64; some are ok but some miss the structures. The loss plot was provided. Nice records of experiment steps. UMAP or T-SNE result was not presented, but the related functions were implemented.
> 2. **Implementation**: Data augmentation implemented. Nice and well structured.
> 3. **Design**: Good
> 4. **Comment**: Basic comments, could have more detailed comments.
> 5. **Difficulty**: Hard
> 
> **Note** Student ID number is required. Is the random seed (manual_seed = 46915474) your ID number?

Hi,

Yes, the random seed is my ID number. Was I required to add comments at the top of each file with my name and student ID?

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. Low image quality.</td><td>-1</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. StyleGAN (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. No embedding style image.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). ON time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is adequate. </td><td>-1</td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-3</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work
